File: mathgenx_app/lib/models/quiz_attempt.dart
class QuizAttempt {
  final String id;
  final String topicId;
  final String subtopicId;
  final String questionFormat;
  final String masteryLevel;
  final String learningStyle;
  final int totalQuestions;
  final int correctAnswers;
  final int incorrectAnswers;
  final double scorePercentage;
  final int marksObtained;
  final int expGained;
  final int totalMarks;
  final DateTime startedAt;
  final DateTime completedAt;
  final int timeSpentSeconds;

  const QuizAttempt({
    required this.id,
    required this.topicId,
    required this.subtopicId,
    required this.questionFormat,
    required this.masteryLevel,
    required this.learningStyle,
    required this.totalQuestions,
    required this.correctAnswers,
    required this.incorrectAnswers,
    required this.scorePercentage,
    required this.marksObtained,
    required this.expGained,
    required this.totalMarks,
    required this.startedAt,
    required this.completedAt,
    required this.timeSpentSeconds,
  });

  factory QuizAttempt.fromJson(Map<String, dynamic> json) => QuizAttempt(
    id: json['id'] as String,
    topicId: json['topic_id'] as String,
    subtopicId: json['subtopic_id'] as String,
    questionFormat: json['question_format'] as String,
    masteryLevel: json['mastery_level'] as String,
    learningStyle: json['learning_style'] as String,
    totalQuestions: (json['total_questions'] as num).toInt(),
    correctAnswers: (json['correct_answers'] as num).toInt(),
    incorrectAnswers: (json['incorrect_answers'] as num).toInt(),
    scorePercentage: double.parse(json['score_percentage'].toString()),
    marksObtained: (json['marks_obtained'] as num).toInt(),
    expGained: (json['exp_gained'] as num).toInt(),
    totalMarks: (json['total_marks'] as num).toInt(),
    startedAt: DateTime.parse(json['started_at'] as String),
    completedAt: DateTime.parse(json['completed_at'] as String),
    timeSpentSeconds: (json['time_spent_seconds'] as num).toInt(),
  );
}

File: mathgenx_backend/app/Models/QuizAttempt.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Concerns\HasUuids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class QuizAttempt extends Model
{
    use HasUuids;

    protected $table = 'quiz_attempts';

    protected $fillable = [
        'user_id',
        'topic_id',
        'subtopic_id',
        'question_format',
        'language',
        'mastery_level',
        'learning_style',
        'total_questions',
        'correct_answers',
        'incorrect_answers',
        'score_percentage',
        'total_marks',
        'marks_obtained',
        'exp_gained',
        'started_at',
        'completed_at',
        'time_spent_seconds',
        'quiz_parameters',
        'source',
    ];

    protected $casts = [
        'started_at' => 'datetime',
        'completed_at' => 'datetime',
        'quiz_parameters' => 'array',
        'score_percentage' => 'decimal:2',
    ];

    // Relationships
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function responses(): HasMany
    {
        return $this->hasMany(QuizResponse::class, 'quiz_attempt_id');
    }

    // Helper methods
    public function calculateScorePercentage(): float
    {
        if ($this->total_questions === 0) {
            return 0;
        }
        return ($this->correct_answers / $this->total_questions) * 100;
    }

    public function getTimeSpentInMinutes(): float
    {
        return $this->time_spent_seconds ? ($this->time_spent_seconds / 60) : 0;
    }
}